### ISSUE TITLE
Pixel wise segmentation: Weight transfer Keras API changes

### ISSUE THREAD

**Issue by ahundt** [jhu-lcsr/costar_plan#357]
UPDATE: The key feature where weights of different dimensions load by name already exists in keras master! This task will be put on hold.

Weights that create single labels per image can be transferred for effective use in per-pixel segmentation tasks. This is demonstrated in the [Keras-FCN](https://github.com/aurora95/Keras-FCN) repository. 

However, even if we construct an otherwise identical model that is simply adapted for pixel-wise tasks they can't be loaded directly with Keras. This is because Keras checks the shape of each set of weights it loads, then exits reporting an error if it does not match.

Weights need to be reshaped, and an example of this being done can be found in:

- Keras_FCN [transfer_FCN.py](https://github.com/aurora95/Keras-FCN/blob/master/utils/transfer_FCN.py)

update the following keras model loading API functions:
 - [x] [load_weights()](https://github.com/keras-team/keras/blob/0611d80cd05c1c199f4c41aae523bf8b988adf24/keras/engine/topology.py#L2600)
 - [x] [def load_weights_from_hdf5_group(f, layers):](https://github.com/keras-team/keras/blob/0611d80cd05c1c199f4c41aae523bf8b988adf24/keras/engine/topology.py#L3093)
 - [x] [def load_weights_from_hdf5_group_by_name(f, layers, skip_mismatch=False):](https://github.com/keras-team/keras/blob/0611d80cd05c1c199f4c41aae523bf8b988adf24/keras/engine/topology.py#L3161)
- [x] [preprocess_weights_for_loading()](https://github.com/keras-team/keras/blob/0611d80cd05c1c199f4c41aae523bf8b988adf24/keras/engine/topology.py#L2886) may or may not also need to be modified.
- [x] [K.batch_set_value()](https://github.com/keras-team/keras/blob/0611d80cd05c1c199f4c41aae523bf8b988adf24/keras/backend/tensorflow_backend.py#L2331) will also need to be updated in each backend.
    - tensorflow_backend.py
    - theano_backend.py
    - cntk_backend.py

1. add a parameter `reshape=False`, which defaults to false
2. If `reshape` is set to `True` compare each layer's weights' array size (total number of entries in the numpy array) to those from the hdf5 weights file (also a numpy array).
    - when size is different report the same error as before or an improved version.
3.  When the array sizes in step 2 are equal, reshape the weights to match the model weights' shape, so they will be imported correctly in the new format.

- [x] update the keras unit test to cover this case
- [x] consider an additional `ignore_missing_layers` option.
